metapool/sample_sheet.py

```diff
@@ -388,6 +353,145 @@ def merge(self, sheets):
+
+            # if a column named 'index' is present in table, assume it is a
+            # numeric index and not a sequence of bases, which is required in
+            # the output. Assume the column that will become 'index' is
+            # defined in remapper.
+            if 'index' in set(out.columns):
+                out.drop(columns=['index'], inplace=True)
+
+            # if an alternate form of a column name defined in
+            # _KL_SAMPLE_SHEET_COLUMN_ALTS is found in table, assume it should
+            # be renamed to its proper form and be included in the output e.g.:
+            # 'sample_plate' -> 'Sample_Plate'.
+
+            # assume keys in _KL_SAMPLE_SHEET_COLUMN_ALTS do not overlap w/
+            # remapper (they currently do not). Define the full set of
+            # potential columns to rename in table.
+
+            # new syntax in 3.9 allows us to merge two dicts together w/OR.
+            remapper = KLSampleSheet.column_alts | self.remapper
+            out.rename(remapper, axis=1, inplace=True)
+
+            # out may contain additional columns that aren't allowed in the
+            # [Data] section of a sample-sheet e.g.: 'Extraction Kit Lot'.
+            # There may also be required columns that aren't defined in out.
+            subset = list(
+                set(KLSampleSheet._KL_SAMPLE_SHEET_DATA_COLUMNS) & set(
+                    out.columns))
+            out = out[subset]
+
+        # append the new 'Well_description' column, now that alternates have
+        # been removed and non-essential columns have been dropped.
+        out['Well_description'] = well_description
+
+        for column in KLSampleSheet._KL_SAMPLE_SHEET_DATA_COLUMNS:
+            if column not in out.columns:
+                warnings.warn('The column %s in the sample sheet is empty' %
+                              column)
+                out[column] = ''
+
+        return out
+
+    def _add_data_to_sheet(self, table, sequencer, lanes, assay, strict=True):
+        table = self._remap_table(table, strict)
+        if assay != _AMPLICON:
+            table['index2'] = sequencer_i5_index(sequencer, table['index2'])
+
+            self.Bioinformatics['BarcodesAreRC'] = str(
+                sequencer in REVCOMP_SEQUENCERS)
+
+        for lane in lanes:
+            for sample in table.to_dict(orient='records'):
+                sample['Lane'] = lane
+                self.add_sample(sample_sheet.Sample(sample))
+
+
+class AmpliconSampleSheet(KLSampleSheet):
+    def __init__(self, path=None):
+        super().__init__(path=path)
+        # _KL_AMPLICON_REMAPPER
+        self.remapper = {
+            'sample sheet Sample_ID': 'Sample_ID',
+            'Sample': 'Sample_Name',
+            'Project Plate': 'Sample_Plate',
+            'Well': 'well_id_384',
+            'Name': 'I7_Index_ID',
+            'Golay Barcode': 'index',
+            'Project Name': 'Sample_Project',
+        }
```

Comment:
I have a very, very strong preference for assigning string literals to a variable once and and using the variable throughout instead of the string, due to having been bitten a number of times by having to change, e.g., a string key in 50 places ... maybe some time in the future we can do that with these strings, most of which seem to be used over and over.

Reply:
I totally agree. What I did here was just move the remapping dicts from the global scope to within individual classes during some early refactoring. I did this just to keep the tests working while I keep refactoring.